components/eamxx/src/physics/gw/impl/gw_gwd_compute_tendencies_from_stress_divergence_impl.hpp

```diff
@@ -15,32 +17,121 @@ template<typename S, typename D>
 KOKKOS_FUNCTION
 void Functions<S,D>::gwd_compute_tendencies_from_stress_divergence(
   // Inputs
-  const Int& ncol,
+  const MemberType& team,
+  const Workspace& workspace,
+  const GwCommonInit& init,
   const Int& pver,
   const Int& pgwv,
   const Int& ngwv,
   const bool& do_taper,
-  const Spack& dt,
-  const Spack& effgw,
-  const uview_1d<const Int>& tend_level,
-  const uview_1d<const Spack>& lat,
-  const uview_1d<const Spack>& dpm,
-  const uview_1d<const Spack>& rdpm,
-  const uview_1d<const Spack>& c,
-  const uview_1d<const Spack>& ubm,
-  const uview_1d<const Spack>& t,
-  const uview_1d<const Spack>& nm,
-  const uview_1d<const Spack>& xv,
-  const uview_1d<const Spack>& yv,
+  const Real& dt,
+  const Real& effgw,
+  const Int& tend_level,
+  const Int& max_level,
+  const Real& lat,
+  const uview_1d<const Real>& dpm,
+  const uview_1d<const Real>& rdpm,
+  const uview_1d<const Real>& c,
+  const uview_1d<const Real>& ubm,
+  const uview_1d<const Real>& t,
+  const uview_1d<const Real>& nm,
+  const Real& xv,
+  const Real& yv,
   // Inputs/Outputs
-  const uview_1d<Spack>& tau,
+  const uview_2d<Real>& tau,
+  const uview_2d<Real>& work,
   // Outputs
-  const uview_1d<Spack>& gwut,
-  const uview_1d<Spack>& utgw,
-  const uview_1d<Spack>& vtgw)
+  const uview_2d<Real>& gwut,
+  const uview_1d<Real>& utgw,
+  const uview_1d<Real>& vtgw)
 {
-  // TODO
-  // Note, argument types may need tweaking. Generator is not always able to tell what needs to be packed
+  const Real ptaper = do_taper ? std::cos(lat) : 1.;
+
+  // Force tau at the top of the model to zero, if requested.
+  if (init.tau_0_ubc) {
+    for (size_t i = 0; i < tau.extent(0); ++i) {
+      tau(i,0) = 0.;
+    }
+  }
+
```

Comment:
I think you need a barrier here before reading tau below.

Alternatively, since the loop below is || is ngwv but serial in k, I'm thinking this 0'ing go right before the serial for loop on k.

Reply:
The main loop is over -ngwv:ngwv, but the tau's first dimension is -pgwv:pgwv, so I don't know if I can do it the way you are suggesting. I can put this 0-ing in a small kernel to reduce races.

Reply:
Inspection of the F90 makes me believe ngwv is an argument to subroutines, and the callers pass pgwv as the value of ngwv. Thus, I provisionally concluded ngwv=pgwv.

Reply:
That's another good reason to take some time to determine this. You could do code inspection to see if my code inspection deduced the right thing. You could also ping Walter Hannah to find out what he thinks. Being able to drop one of the two index ranges would be nice.

Reply:
Done